presto-orc/src/main/java/com/facebook/presto/orc/zstd/FiniteStateEntropy.java

```diff
@@ -0,0 +1,168 @@
+
+            UNSAFE.putByte(outputBase, output + 2, symbols[state1]);
+            int numberOfBits2 = numbersOfBits[state1];
+            state1 = (int) (newStates[state1] + peekBits(bitsConsumed, bits, numberOfBits2));
+            bitsConsumed += numberOfBits2;
+
+            UNSAFE.putByte(outputBase, output + 3, symbols[state2]);
+            int numberOfBits3 = numbersOfBits[state2];
+            state2 = (int) (newStates[state2] + peekBits(bitsConsumed, bits, numberOfBits3));
+            bitsConsumed += numberOfBits3;
+
+            output += SIZE_OF_INT;
+
+            loader = new BitStream.Loader(inputBase, input, currentAddress, bits, bitsConsumed);
+            boolean done = loader.load();
+            bitsConsumed = loader.getBitsConsumed();
+            bits = loader.getBits();
+            currentAddress = loader.getCurrentAddress();
+            if (done) {
+                break;
+            }
+        }
+
+        while (true) {
+            UNSAFE.putByte(outputBase, output++, symbols[state1]);
+            int numberOfBits = numbersOfBits[state1];
+            state1 = (int) (newStates[state1] + peekBits(bitsConsumed, bits, numberOfBits));
+            bitsConsumed += numberOfBits;
+
+            loader = new BitStream.Loader(inputBase, input, currentAddress, bits, bitsConsumed);
+            loader.load();
+            bitsConsumed = loader.getBitsConsumed();
+            bits = loader.getBits();
+            currentAddress = loader.getCurrentAddress();
+
+            if (loader.isOverflow()) {
+                UNSAFE.putByte(outputBase, output++, symbols[state2]);
+                break;
+            }
+
+            UNSAFE.putByte(outputBase, output++, symbols[state2]);
+            int numberOfBits1 = numbersOfBits[state2];
+            state2 = (int) (newStates[state2] + peekBits(bitsConsumed, bits, numberOfBits1));
+            bitsConsumed += numberOfBits1;
+
+            loader = new BitStream.Loader(inputBase, input, currentAddress, bits, bitsConsumed);
+            loader.load();
+            bitsConsumed = loader.getBitsConsumed();
+            bits = loader.getBits();
+            currentAddress = loader.getCurrentAddress();
+
+            if (loader.isOverflow()) {
+                UNSAFE.putByte(outputBase, output++, symbols[state1]);
+                break;
+            }
+        }
+
+        return (int) (output - outputAddress);
+    }
+
+    public static final class Table
+    {
+        int tableLog;
+        final int[] newState;
+        final byte[] symbol;
+        final byte[] numberOfBits;
+
+        public Table(int maxLog)
+        {
+            int size = 1 << maxLog;
+            newState = new int[size];
+            symbol = new byte[size];
+            numberOfBits = new byte[size];
+        }
+
+        public Table(int tableLog, int[] newState, byte[] symbol, byte[] numberOfBits)
+        {
+            this.tableLog = tableLog; // TODO: can be inferred from array lengths?
```

Comment:
Yes, they should be the same size. I believe they have to be powers of 2, so they need to match tableLog.